Lib/subprocess.py

```diff
@@ -1456,7 +1460,10 @@ def _posix_spawn(self, args, executable, env, restore_signals):
                         sigset.append(signum)
                 kwargs['setsigdef'] = sigset
 
-            self.pid = os.posix_spawn(executable, args, env, **kwargs)
+            if _HAVE_POSIX_SPAWNP:
+                self.pid = os.posix_spawnp(executable, args, env, **kwargs)
```

Comment:
Please keep in mind that `_posixsubprocess` is likely to use a different algorithm for searching in `PATH` compared to `posix_spawnp` implementations (and, moreover, `posix_spawnp` implementations differ between libcs and even between versions of the same libc). For example, musl uses [`execvpe`](https://git.musl-libc.org/cgit/musl/tree/src/process/execvp.c#n42) as a "backend" for `posix_spawnp`, which skips only 3 kinds of errors, while [`_posixsubprocess` skips all errors](https://github.com/python/cpython/blob/07858894689047c77f9c12ddc061d30681368d19/Modules/_posixsubprocess.c#L518). Also, `EACCES` is treated in a special way by both glibc and musl.

Also, the default `PATH` if the environment variable is not set is likely to differ (e.g., glibc uses `/bin:/usr/bin`, musl uses `/usr/local/bin:/bin:/usr/bin`, and Python uses `:/bin:/usr/bin`).

Reply:
"... glibc uses /bin:/usr/bin ... and Python uses :/bin:/usr/bin"

Oh. ":" at the start of posixpath.defpath is surprising. Is it a bug? What's the point of putting an empty string in the list of default search paths? It doesn't help to locate an executable...

Reply:
It's not a bug: [empty entries in PATH mean "the current directory"](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08_03). See also NOTES in [man execvp](http://man7.org/linux/man-pages/man3/execvp.3.html) about glibc dropping this empty entry.

Reply:
I created https://bugs.python.org/issue35755 to propose to remove it :-)